src/target/riscv/riscv-013.c

```diff
@@ -872,17 +877,27 @@ static int execute_abstract_command(struct target *target, uint32_t command)
 		}
 	}
 
-	if (dm_write_exec(target, DM_COMMAND, command, false) != ERROR_OK)
+	if (dm_write_exec(target, DM_COMMAND, command, false /* ensure success */) != ERROR_OK)
 		return ERROR_FAIL;
 
-	uint32_t abstractcs = 0;
-	int result = wait_for_idle(target, &abstractcs);
-
-	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
-	if (info->cmderr != 0 || result != ERROR_OK) {
-		LOG_TARGET_DEBUG(target, "command 0x%x failed; abstractcs=0x%x", command, abstractcs);
-		/* Clear the error. */
-		dm_write(target, DM_ABSTRACTCS, DM_ABSTRACTCS_CMDERR);
+	uint32_t abstractcs;
+	int wait_result = wait_for_idle(target, &abstractcs);
+	if (wait_result != ERROR_OK) {
+		/* TODO: can we recover from this? */
+		if (wait_result == ERROR_TIMEOUT_REACHED)
+			LOG_TARGET_DEBUG(target, "command 0x%" PRIx32 " failed (timeout)", command);
+		else
+			LOG_TARGET_DEBUG(target, "command 0x%" PRIx32 " failed (unknown fatal error %d)", command, wait_result);
+		return wait_result;
+	}
+	*cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	if (*cmderr != CMDERR_NONE) {
+		LOG_TARGET_DEBUG(target, "command 0x%" PRIx32 " failed; abstractcs=0x%" PRIx32 "",
+			command, abstractcs);
+		/* Attempt to clear the error. */
+		/* TODO: can we add a more substantial recovery if the clear operation fails ? */
+		if (dm_write(target, DM_ABSTRACTCS, DM_ABSTRACTCS_CMDERR) != ERROR_OK)
+			LOG_TARGET_ERROR(target, "could not clear abstractcs error");
```

Comment:
I would recommend to remove this TODO.

If a simple write to a debug module register fails (in this case write to abstractcs), then the target is seriously broken already. I'm afraid there is nothing more we could do, apart from printing an error. 

```suggestion
		if (dm_write(target, DM_ABSTRACTCS, DM_ABSTRACTCS_CMDERR) != ERROR_OK)
			LOG_TARGET_ERROR(target, "could not clear abstractcs error");
```

Reply:
> I'm afraid there is nothing more we could do, apart from printing an error. 

The TODO is to investigate if we have any options to recover somehow. For example @en-sc had a suggestion (during an internal review) to just drop a marker that target target was examined to enforce debugger to restore the state or give a user a reasonable feedback. **maybe** we could also reset the DM. I mean there may be some crazy variant that will make a recovery possible.

Reply:
Thank you for explanation.

I believe it is a more general question about what OpenOCD should do (if anything at all) when it encounters an evidently broken target.

An argument could be made to just report this error to the user, and let the user deal with it (which would be my preference in such situations).

Anyway, let's leave the TODO in place as you originally suggested.